packages/core/src/editor/BlockNoteExtension.ts

```diff
@@ -123,78 +121,104 @@ type InputRule = {
+    | Record<string, any>
+    | undefined,
+> {
+  options: Options;
+  editor: BlockNoteEditor<any, any, any>;
+}
 
 // a type that maps the extension key to the return type of the extension factory
-export type ExtensionMap<
-  T extends ReadonlyArray<ExtensionFactory | undefined>,
-> = {
-  [K in ExtractExtensionKey<T[number]>]: ExtractExtensionByKey<
-    Exclude<T[number], undefined>,
-    K
-  >;
+export type ExtensionMap<T extends ReadonlyArray<ExtensionFactoryInstance>> = {
+  [K in T[number] extends ExtensionFactoryInstance<infer Ext>
+    ? Ext["key"]
+    : never]: T[number] extends ExtensionFactoryInstance<infer Ext>
+    ? Ext
+    : never;
 };
 
 /**
- * Helper function to create a BlockNote extension.
- * Can accept either an ExtensionFactory (function) or an Extension (object).
- * If an Extension is provided, it will be wrapped in a factory function.
+ * This is a type that represents the function which will actually create the extension.
+ * It requires the editor instance to be passed in, but will already have the options applied automatically.
+ *
+ * @note Only the BlockNoteEditor should instantiate this function, not the user. Look at {@link createExtension} for user-facing functions.
  */
-export function createExtension<const T extends ExtensionFactory>(ext: T): T;
-export function createExtension<const Key extends string, const State = any>(
-  ext: Extension<State, Key>,
-): () => Extension<State, Key>;
-export function createExtension<
-  const T extends ExtensionFactory | Extension | undefined = any,
->(
-  extension: T,
-): T extends Extension
-  ? () => T
-  : T extends ExtensionFactory
-    ? T
-    : () => undefined {
-  if (typeof extension === "function") {
-    return extension as any;
-  }
+export type ExtensionFactoryInstance<
+  Ext extends Extension<any, any> = Extension<any, any>,
+> = (ctx: Omit<ExtensionOptions<any>, "options">) => Ext;
 
-  if (typeof extension === "object" && "key" in extension) {
-    return (() => extension) as any;
-  }
+/**
+ * This is the return type of the {@link createExtension} function.
+ * It is a function that can be invoked with the extension's options to create a new extension factory.
+ */
+export type ExtensionFactory<
+  State = any,
+  Key extends string = string,
+  Factory extends (ctx: any) => Extension<State, Key> = (
+    ctx: ExtensionOptions<any>,
+  ) => Extension<State, Key>,
+> =
+  Parameters<Factory>[0] extends ExtensionOptions<infer Options>
+    ? undefined extends Options
+      ? (
+          options?: Exclude<Options, undefined>,
+        ) => ExtensionFactoryInstance<ReturnType<Factory>>
+      : (options: Options) => ExtensionFactoryInstance<ReturnType<Factory>>
+    : () => ExtensionFactoryInstance<ReturnType<Factory>>;
 
-  if (!extension) {
-    return (() => undefined) as any;
+/**
+ * Constructs a BlockNote {@link ExtensionFactory} from a factory function or object
+ */
+export function createExtension<
```

Comment:
I think it could help (me at least..) to make sense of all overloads if every overload has a tiny 1-line example of what the usage looks like (`// this overload is for createExtension({...})`